### ISSUE TITLE
Clarify input/output resource concurrency limitations

### ISSUE THREAD

**Issue by iluminae** [redpanda-data/connect#706]
Using the resource type makes the pipeline _much_ more readable and fixes all of the _terrible_ metric names: (`myingestionpipeline_output_broker_outputs_0_switch_0_output_sent` eek [using prometheus if that was not obvious]).

However, using `resource` _kills_ the concurrency patterns we have available to us. I know this is said in the first paragraph of [the resources document](https://www.benthos.dev/docs/configuration/resources):
```
Only one instance of each named resource is created ...
``` 
However, it was not clear to me from that that _I could not_ use them effectively with brokered inputs/outputs and many-threaded pipelines to achieve higher throughput.

When I had config:
```yaml
output:
  broker:
    copies: 10
    pattern: greedy
    outputs:
    - resource: mything
resources:
  outputs:
    mything:
      type: mynetworkthing
```
It would spin up 1x`mynetworkthing` and really ignore completely that the output was specified to run 10 of them - they were all the _same thing_, and therefore would be blocking each other using the same transactionChan. Using resources here changed my metrics from `myingestionpipeline_output_broker_outputs_0_switch_0_output_sent` to `myingestionpipeline_resource_output_mything_sent` which is much better but obviously reduced my throughput dramatically.

I believe the same goes for the input brokers, and the processors. If you have `threads: 10` but you have a processor that is _not pure_ (meaning it does something outside of the message, like a network request) then really you would only get 1x of them initialized and use the same one for all 10 threads (for pure processors I dont think this is a problem, but for the ones that eg: use a different conn per instance you will use the same conn)

Anyway I understand now that resources=1 of them - so maybe it just needs to be more strictly documented? Aside of that - I can use the `path_mapping` feature of the prometheus metric output to change my stats... and that is _fine_.

**Comment by Jeffail** [comment]
Hey @iluminae, for processors it should still be fine even with networked components as they should be capable of supporting multiple in-flight invocations. However, the inputs and outputs are sort of a special case as they're much more stateful, I can add clearer documentation to them specifically as they're the most likely pain point.

It might also be worth revisiting the output itself as many of them now have a `max_in_flight` field, where increasing that can often have a better performance impact than using a `greedy` broker.

For metrics labelling I am extremely close to implementing https://github.com/Jeffail/benthos/issues/510, which would still allow you to use resources, but you wouldn't be forced to use it just for improving your metric names. My focus right now is getting v3.43.0 out the door, which comes with a lot of fundamental improvements to Benthos internals, once that's done I'll be unblocked to finish up labels and the plugins v2 APIs.

**Comment by Jeffail** [comment]
Also it would make sense for Benthos to automatically detect a broker with `copies: N` when the child outputs are resources and provide a linting error.

**Comment by iluminae** [comment]
Documentation is fine, just wanted to bring it up as it caused me quite a few slower-than-expected pipelines.

**Comment by twmb** [comment]
Closing as part of a backlog cleanup. The underlying question was answered in the comments, and metric label support has since been added (removing the main motivation for using resources just for better metric names). Documentation has also been substantially updated. If there are specific gaps, feel free to open a new issue.
